lib/Transforms/SecretWithMgmt/SecretWithMgmtBGV.cpp

```diff
@@ -0,0 +1,385 @@
+    solver.load<dataflow::SparseConstantPropagation>();
+    solver.load<SecretnessAnalysis>();
+    if (failed(solver.initializeAndRun(getOperation()))) {
+      getOperation()->emitOpError() << "Failed to run the analysis.\n";
+      signalPassFailure();
+      return;
+    }
+
+    auto getSecretness = [&](Value value) {
+      const auto *lattice = solver.lookupState<SecretnessLattice>(value);
+      if (!lattice) {
+        // newly created value does not have lattice
+        return true;
+      }
+      auto &secretness =
+          solver.lookupState<SecretnessLattice>(value)->getValue();
+      if (secretness.isInitialized()) {
+        return secretness.getSecretness();
+      }
+      // if not initialized assume secret
+      return true;
+    };
+
+    DenseMap<Value, int> levelMap;
+    DenseMap<Value, bool> mulMap;  // check a Value is a mul result or not
+
+    OpBuilder b(&getContext());
+    getOperation()->walk<WalkOrder::PreOrder>([&](secret::GenericOp genericOp) {
+      for (auto blockArg : genericOp.getBody()->getArguments()) {
+        levelMap[blockArg] = 0;
+        mulMap[blockArg] = false;
+      }
+
+      genericOp.getBody()->walk<WalkOrder::PreOrder>([&](Operation *op) {
+        auto levelResult = 0;
+        for (auto operand : op->getOperands()) {
+          if (levelMap.count(operand) == 0) {
+            continue;
+          }
+          auto levelOperand = levelMap.at(operand);
+          levelResult = std::max(levelResult, levelOperand);
+        }
+
+        bool operandsMul = false;
+        for (auto operand : op->getOperands()) {
+          if (mulMap.count(operand) == 0) {
+            continue;
+          }
+          operandsMul |= mulMap.at(operand);
+        }
+
+        llvm::TypeSwitch<Operation &>(*op)
+            .Case<arith::MulIOp, secret::YieldOp, tensor::ExtractOp>(
+                [&](auto mulOp) {
+                  // mod reduced
+                  if ((isa<arith::MulIOp>(mulOp) ||
+                       isa<tensor::ExtractOp>(mulOp)) &&
+                      (includeFirst || operandsMul)) {
+                    levelResult += 1;
+                  }
+                  if (isa<secret::YieldOp>(mulOp) && operandsMul) {
+                    // mod reduce before yield if muled
+                    levelResult += 1;
+                  }
+                  // avoid yield op
+                  if (mulOp->getNumResults() != 0) {
+                    levelMap[mulOp->getResult(0)] = levelResult;
+                    mulMap[mulOp->getResult(0)] = true;
+                  }
+
+                  for (auto operand : mulOp->getOperands()) {
+                    auto secretness = getSecretness(operand);
+                    if (!secretness) {
+                      continue;
+                    }
+                    b.setInsertionPoint(mulOp);
+                    Value managed = operand;
+                    for (auto i = 0; i != levelResult - levelMap.at(operand);
```

Comment:
```suggestion
                    for (auto i = 0; i < levelResult - levelMap.at(operand);
```

nit: protect against a potential infinite loop